src/plot_api/template_api.js

```diff
@@ -269,3 +269,187 @@ function getNextPath(parent, key, path) {
 
     return nextPath;
 }
+
+/**
+ * validateTemplate: Test for consistency between the given figure and
+ * a template, either already included in the figure or given separately.
+ * Note that not every issue we identify here is necessarily a problem,
+ * it depends on what you're using the template for.
+ *
+ * @param {object|DOM element} figure: the plot, with {data, layout} members,
+ *     to test the template against
+ * @param {Optional(object)} template: the template, with its own {data, layout},
+ *     to test. If omitted, we will look for a template already attached as the
+ *     plot's `layout.template` attribute.
+ *
+ * @returns {array} array of error objects each containing:
+ *  - {string} code
+ *      error code ('missing', 'unused', 'reused', 'noLayout', 'noData')
+ *  - {string} msg
+ *      a full readable description of the issue.
+ */
+exports.validateTemplate = function(figureIn, template) {
+    var figure = Lib.extendDeep({}, {
+        _context: dfltConfig,
+        data: figureIn.data,
+        layout: figureIn.layout
+    });
+    var layout = figure.layout || {};
+    if(!isPlainObject(template)) template = layout.template || {};
+    var layoutTemplate = template.layout;
+    var dataTemplate = template.data;
+    var errorList = [];
+
+    figure.layout = layout;
+    figure.layout.template = template;
+    Plots.supplyDefaults(figure);
+
+    var fullLayout = figure._fullLayout;
+    var fullData = figure._fullData;
+
+    if(!isPlainObject(layoutTemplate)) {
+        errorList.push({code: 'layout'});
+    }
+    else {
+        // TODO: any need to look deeper than the first level of layout?
+        // I don't think so, that gets all the subplot types which should be
+        // sufficient.
```

Comment:
What about `xaxis.rangeslider.yaxis2` from https://github.com/plotly/plotly.js/pull/2364?

Reply:
Hmm, good point. You're right, though it's a lot of complication (for example for `xaxis.rangeslider.yaxis` we'd have to look for `xaxis<N>.rangeslider.yaxis<M>` for any combination of `<N>` and `<M>` and only generate an error if we don't find *any* of them. And the only time this would alert you that the template isn't appropriate is if the template has a `xaxis<N>.rangeslider.yaxis<M>` but does NOT have `yaxis<M>` - ie your template had opinions about how that y axis should work inside the rangeslider, but left the y axis itself untouched. This is of course possible, just seems very unlikely.

Reply:
> This is of course possible, just seems very unlikely.

Right. I don't this needs special handling. A comment about this would suffice.

Reply:
actually wasn't all that hard - recursed into layout in https://github.com/plotly/plotly.js/pull/2761/commits/15931cf17ebb77baa108b85d80237bfd40cec700